### ISSUE TITLE
Rust 1.26.0 fails to build via homebrew on OS X 10.10.5

### ISSUE THREAD

**Issue by cilia** [rust-lang/rust#50776]
Trying to compile and install Rust 1.26.0 on OSX 10.10.5 using Homebrew. Tried a couple of times, both failed with the same msgs, which seemed the failure happened at `cargo` or `std`. Please see msgs below

```
Compiling panic_abort v0.0.0 (file:///private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/src/libpanic_abort)
Compiling alloc v0.0.0 (file:///private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/src/liballoc)
Compiling panic_unwind v0.0.0 (file:///private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/src/libpanic_unwind)
Compiling alloc_system v0.0.0 (file:///private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/src/liballoc_system)
error: Could not compile `std`.
```

```
Caused by:
process didn't exit successfully: `/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/bootstrap/debug/rustc --crate- 
name std libstd/lib.rs --error-format json --crate-type dylib --crate-type rlib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -- 
cfg feature="alloc_jemalloc" --cfg feature="backtrace" --cfg feature="jemalloc" --cfg feature="panic-unwind" --cfg 
feature="panic_unwind" -C metadata=a297a638e0984b03 -C extra-filename=-a297a638e0984b03 --out-dir /private/tmp/rust- 
20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target 
x86_64-apple-darwin -L dependency=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple- 
darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0- 
src/build/x86_64-apple-darwin/stage1-std/release/deps --extern rustc_asan=/private/tmp/rust-20180514-70415-ya9x99/rustc- 
1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/librustc_asan-f4f6bcd504eef548.rlib --extern 
panic_abort=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple- 
darwin/release/deps/libpanic_abort-c5809d2b73f594e5.rlib --extern alloc_system=/private/tmp/rust-20180514-70415-ya9x99/rustc- 
1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/liballoc_system-6f7baa29581dba15.rlib -- 
extern compiler_builtins=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1- 
std/x86_64-apple-darwin/release/deps/libcompiler_builtins-bd73167a8925a884.rlib --extern core=/private/tmp/rust-20180514- 
70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore- 
25b6d1996af3ba6a.rlib --extern alloc=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple- 
darwin/stage1-std/x86_64-apple-darwin/release/deps/liballoc-bfe57e517b27a55f.rlib --extern alloc_jemalloc=/private/tmp/rust- 
20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple- 
darwin/release/deps/liballoc_jemalloc-03d6ea9f31c524d2.rlib --extern rustc_tsan=/private/tmp/rust-20180514-70415-ya9x99/rustc- 
1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/librustc_tsan-a8408b80f28a7477.rlib --extern 
libc=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple- 
darwin/release/deps/liblibc-4fd7000939cb1dfa.rlib --extern std_unicode=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0- 
src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libstd_unicode-bb9e3d02f8a36008.rlib --extern 
panic_unwind=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple- 
darwin/release/deps/libpanic_unwind-5f5b955434cfdf44.rlib --extern unwind=/private/tmp/rust-20180514-70415-ya9x99/rustc- 
1.26.0-src/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libunwind-36f333beda82c695.rlib -L 
native=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple-darwin/native/libbacktrace/.libs -l 
static=backtrace -l System -l resolv -L native=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple- 
darwin/native/jemalloc/lib -L native=/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple- 
darwin/stage1-std/x86_64-apple-darwin/release/build/compiler_builtins-aa56d6d2d94a5702/out` (signal: 11, SIGSEGV: invalid 
memory reference)
command did not execute successfully: "/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/x86_64-apple- 
darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "--release" "--features" "panic-unwind jemalloc backtrace" "- - 
manifest-path" "/private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101
thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1064:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /private/tmp/rust-20180514-70415-ya9x99/rustc-1.26.0-src/build/bootstrap/debug/bootstrap build
Build completed unsuccessfully in 0:24:31
make: *** [all] Error 1
```

Reddit discussion [here](https://www.reddit.com/r/rust/comments/8jg44m/rust_1260_compilation_fails_on_os_x_10105/)

**Comment by cilia** [comment]
Any update?

**Comment by Aaronepower** [comment]
Triage; @cilia Hello, are you still having this issue on the latest release?

**Comment by cilia** [comment]
@Aaronepower Yes, just tried 1.29.1, still failing. Please see the error msg below

`==> Upgrading rust`
`==> Downloading https://static.rust-lang.org/dist/rustc-1.29.1-src.tar.gz
######################################################################## 100.0%`
`==> ./configure --prefix=/usr/local/Cellar/rust/1.29.1 --release-channel=stable`
`==> make
Last 15 lines from /Users/cilia/Library/Logs/Homebrew/rust/02.make:`
   `Compiling humantime v1.1.1`
   `Compiling smallvec v0.6.3`
`error: Could not compile rustc_platform_intrinsics`.

`Caused by:`
  `process didn't exit successfully: /private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/bootstrap/debug/rustc --crate-name rustc_platform_intrinsics librustc_platform_intrinsics/lib.rs --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=72a617c6fad69483 -C extra-filename=-72a617c6fad69483 --out-dir /private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/x86_64-apple-darwin/stage0-rustc/x86_64-apple-darwin/release/deps -L dependency=/private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/x86_64-apple-darwin/stage0-rustc/release/deps (signal: 6, SIGABRT: process abort signal)
warning: build failed, waiting for other jobs to finish...`
`error: build failed`
`command did not execute successfully: "/private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "8" "--release" "--features" " jemalloc" "--manifest-path" "/private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/src/rustc/Cargo.toml" "--message-format" "json"`
`expected success, got: exit code: 101`
`thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1119:9`
`note: Run with RUST_BACKTRACE=1 for a backtrace.`
`failed to run: /private/tmp/rust-20181002-68746-1e4xso6/rustc-1.29.1-src/build/bootstrap/debug/bootstrap build`
`Build completed unsuccessfully in 0:03:55`
`make: *** [all] Error 1`

**Comment by pnkfelix** [comment]
discussed at compiler team meeting. I don't know if anyone on the team claims responsibility for helping with homebrew (or relatively old version of OS X).

Assigning to self to investigate.

**Comment by pnkfelix** [comment]
(marking as P-high to remind self to investigate circumstances of bug. But the particulars: Homebrew + old Mac OS X version could well lead to P-medium triaging.)

**Comment by pnkfelix** [comment]
while I wait for some trial runs to go forward on my machine, here is a potentially relevant question:

@cilia , what version of XCode do you have installed on your system? (I think running `xcodebuild --version` will tell me what I need to know here.)

**Comment by pnkfelix** [comment]
In the meantime, I have successfully installed rust 1.30 from homebrew on Mac OS X 10.13 "High Sierra"; but I'm pretty sure that install was using prebuilt binaries. I'm trying again now with the `--build-from-source` option; we'll see where that goes.

In parallel, I have briefly experimented with trying to set up a VMWare image with Mac OS X 10.10 "Yosemite"; but it seems to be a bit of a pain to get such a configuration working. (It appears one might need to first download and install "Mountain Lion" on the VM, and then do an upgrade? Except the App Store won't let me download "Mountain Lion" because its not compatible with *this* laptop...)

(Mac OS X 10.10 "Yosemite" was released Oct 2014, so it *is* pretty old at this point...)

**Comment by pnkfelix** [comment]
I was indeed able to do ` brew install rust --build-from-source` on my laptop running Mac OS X 10.13 "High Sierra"

So this seems like a problem one needs to pull out an old OS to replicate...

**Comment by pnkfelix** [comment]
(I grabbed .dmg images of old installation media, since Apple does keep those available for download, but I was not able to use them with vmware, even after converting them to .iso's in Disk Utility... so, no progress yet in replicating this...)

**Comment by pnkfelix** [comment]
I don't know if its realistic for us to resolve this problem.

Cc @rust-lang/release should we just officially state that we don't actively support building new versions of Rust atop such old versions of Mac OS X? If so, what *should* we say is minimum supported Mac OS X version?

Nominated for discussion amongst T-release team.

**Comment by XAMPPRocky** [comment]
So this is a bit of a complicated question. On the vendor side Apple does not ever announce a "end of life" for any of the major operating system versions, they simply stop updating them. Historically though if we're going by Apple's [security updates], the last three releases, including the current one are supported. So today that would be Catalina (10.15), Mojave (10.14.6), and High Sierra (10.13.6). So for this issue Yosemite (10.10) would definitely fall out of the support window.

Someone else who knows more about our current infra could probably answer to how difficult it is to build and test for older apple platforms.

[security updates]: https://support.apple.com/en-us/HT201222

**Comment by Mark-Simulacrum** [comment]
I am in favor of closing this as won't fix. It seems pretty clear that reproducing it (and so tracking down where the SIGSEGV is coming from) is pretty hard, so we probably can't easily track down the root cause and fix it; furthermore, I'd say that if the bug is strictly that homebrew can't build Rust then that's also likely out of scope for this issue tracker.

I believe we currently support cross-compile targeting "64-bit OSX (10.7+, Lion+)" per the [platform support](https://forge.rust-lang.org/release/platform-support.html) page. But that's not where Rust itself can be built, right?

**Comment by pnkfelix** [comment]
closing as wont fix, following reasoning given by @Mark-Simulacrum

**Comment by frakman1** [comment]
I just wanted to add my experience to this.
I tried to do a `brew upgrade` on my Mac OSX running High Sierra (10.13) and got this rust related error:;

```bash
==> Installing vala dependency: rust
==> ./configure --prefix=/usr/local/Cellar/rust/1.49.0_1 --release-channel=stable
==> make
==> make install
==> ./install.sh --prefix=/private/tmp/rust-20210201-21234-2auhx1/rustc-1.49.0-src/cargobootstrap
==> cargo install /usr/local/Cellar/rust/1.49.0_1 --features curl-sys/force-system-lib-on-osx
Last 15 lines from /Users/frak/Library/Logs/Homebrew/rust/05.cargo:
    |                                                           this method call resolves to `&T`
    |
    = note: cannot satisfy `std::string::String: AsRef<_>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
error: could not compile `cargo`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo v0.50.0 (/private/tmp/rust--cargo-20210201-21234-fgm2gx)`, intermediate artifacts can be found at `/private/tmp/rust--cargo-20210201-21234-fgm2gx/target`

Caused by:
  build failed

Do not report this issue to Homebrew/brew or Homebrew/core!

These open issues may also help:
Rust-dependent formulae on Apple Silicon - upstream issue tracker https://github.com/Homebrew/homebrew-core/issues/68301

Error: You are using macOS 10.13.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.
```

I then installed rust manually via `curl https://sh.rustup.rs -sSf | sh` with default installation option and updated the environment with `source $HOME/.cargo/env` but `brew upgrade` still wanted to install/compile rust again.

**Comment by andrewleek-droplab** [comment]
@frakman1

I have run into the exact same problem as you, although i was trying to upgrade ffmpeg.  What has worked for me was to use the most up-to-date formula i could find that still had a bottle for high-sierra.  Basically i ran `brew edit rust` and then replace the formula with this one: https://raw.githubusercontent.com/Homebrew/homebrew-core/5b4630757e8748165fe69fbefa0fa7949e64a231/Formula/rust.rb

Then i ran `brew install rust` and everything installed fine.

Hope this helps...

**Comment by frakman1** [comment]
@andrewleek-droplab 
Thank you so much! I don't much about brew's backend (what is a bottle, cask, cellar, keg?) so this was quite informative.

**Comment by Klemart3D** [comment]
@andrewleek-droplab answer didn't works for me, still having issues with rust or cargo :(

@frakman1 : https://docs.brew.sh/Formula-Cookbook#homebrew-terminology

**Comment by differance1** [comment]
@andrewleek-droplab -- worked for me, also trying to upgrade ffmpeg !

Had to brew edit formulas from the same repository for cargo-c and rav1e, but other dependencies worked, except had to force the current asciidoc to link (brew link --overwrite asciidoc) after brew upgrade ffmpeg completed.

Really appreciate the tip, gave me a key window into the whole brew getup, have a sense now of how and where I can start trying to hack install processes like brew, an area I'd never gotten into.  :-)

**Comment by differance1** [comment]
@andrewleek-droplab -- worked for me, also trying to upgrade ffmpeg !

Had to brew edit formulas from the same repository for cargo-c and rav1e, but other dependencies worked, except had to force the current asciidoc to link (brew link --overwrite asciidoc) after brew upgrade ffmpeg completed.

Really appreciate the tip, gave me a key window into the whole brew getup, have a sense now of how and where I can start trying to hack install processes like brew, an area I'd never gotten into.  :-)

**Comment by francisuk1989** [comment]
```
==> ./configure --prefix=/usr/local/Cellar/rust/1.54.0_1 --release-channel=stable
==> make

Last 15 lines from /Users/francis/Library/Logs/Homebrew/rust/02.make:
[907/2756] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/CmpInstAnalysis.cpp.o
[908/2756] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/CallGraphSCCPass.cpp.o
[909/2756] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/CallPrinter.cpp.o
[910/2756] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/CaptureTracking.cpp.o
[911/2756] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/CGSCCPassManager.cpp.o
ninja: build stopped: subcommand failed.
thread 'main' panicked at '
command did not execute successfully, got: exit status: 1

build script failed, must exit now', /Users/francis/Library/Caches/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.44/src/lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
	finished in 532.433 seconds
failed to run: /private/tmp/rust-20210905-23742-18ybj5m/rustc-1.54.0-src/build/bootstrap/debug/bootstrap build --stage 2
Build completed unsuccessfully in 0:11:33
make: *** [all] Error 1

```

**Comment by ca13ra1** [comment]
@Jerry1144 I’m having the exact same issue. Can’t seem to find a fix or workaround for it running High Sierra.

**Comment by YaserMb** [comment]
I am facing the same issue. I am installing facebook watchman....while installing watchman it is installing its dependency which is rust....and brew is not able to install rust on high serria

**Comment by battaglia01** [comment]
Is it perhaps possible to install rust from some other source and make Homebrew think it's already installed?

**Comment by YvanQu** [comment]
LMAO I just downloaded the binaries from [https://ffbinaries.com/downloads](url), threw them in the ffmpeg folder in the cellar and ran brew link ffmpeg

**Comment by joaocarmo** [comment]
I was having @Jerry1144's exact same issue, so I followed @skodises' suggestion:

1. Installed **rustup** with `curl https://sh.rustup.rs -sSf | sh`
2. Linked **rust** with `mkdir -p "$(brew --cellar rust)/$(rustc --version | cut -d' ' -f2)" && ln -s "$(dirname $(which rustc))" "$(brew --cellar rust)/$(rustc --version | cut -d' ' -f2)/bin" && brew link rust`
3. Installed **cargo-c** with `cargo install cargo-c`
4. Linked **cargo-c** with `mkdir -p "$(brew --cellar cargo-c)/$(cargo-capi --version | awk '{print $2}' | cut -d'+' -f1)/bin" && cargobins=("cargo-capi" "cargo-cbuild" "cargo-cinstall" "cargo-ctest"); for cargobin in ${cargobins[@]}; do ln -s "$(which $cargobin)" "$(brew --cellar cargo-c)/$(cargo-capi --version | awk '{print $2}' | cut -d'+' -f1)/bin/$cargobin"; done && brew link cargo-c`

Seems to _kind of_ work, but every cargo dependency now has to be installed and linked in a similar way or we get this error:

```txt
==> cargo install
Last 15 lines from /Users/{user}/Library/Logs/Homebrew/{dependency}/01.cargo:
2021-11-14 15:14:43 +0000

cargo
install
--locked
--root
/usr/local/Cellar/{dependency}/{version}
--path
.

error: no override and no default toolchain set
```

**Comment by bvahdat** [comment]
I was able to successfully install the latest rust version `1.56.1` on macOS Heigh Sierra thanks to [this PR](https://github.com/Homebrew/homebrew-core/pull/89988) which has been merged yesterday.

**Comment by catap** [comment]
FYI: with some work for last week I've successfully build `1.57.0` on 10.7+ via MacPorts: https://ports.macports.org/port/rust/details/

**Comment by catap** [comment]
@pnkfelix is it possible to build (somehow) `rustc` with statically linked  rust's libstd and librustc_driver`?

**Comment by frakman1** [comment]
I was able to get brew to install rust version 1.56.1 on my High Sierra MacOS using the link from bvahdat. 
You basically just have to have the latest version of `Formula/rust.rb`.

```
$rustc --version
rustc 1.56.1

$sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G65
```

**Comment by bbgdzxng1** [comment]
Big thanks to @carlocab.  After a clean homebrew uninstall and reinstall, rust finally installed nicely on High Sierra without editing any files.

**Comment by sanxiyn** [comment]
Closing, presuming fixed on Homebrew side.
